osvr/RenderKit/RenderManagerD3D11ATW.h

```diff
@@ -528,62 +514,68 @@ namespace osvr {
-                          setDoingOkay(false);
-                          return false;
-                      }
-
                       // Record the place to copy incoming textures to.
                       newInfo.textureCopy = textureCopy;
                     }
 
-                    // We can't use the render thread's ID3D11RenderTargetView. Create one from
-                    // the ATW's ID3D11Texture2D handle.
-                    newInfo.atwBuffer.D3D11 = new osvr::renderkit::RenderBufferD3D11();
-                    newInfo.atwBuffer.D3D11->colorBuffer = texture2D;
+                    // If we made a new texture, we get our info from it.  Otherwise,
+                    // from the original texture.
+                    ID3D11Texture2D *textureToMap = buffers[i].D3D11->colorBuffer;
+                    if (newInfo.textureCopy) {
+                      textureToMap = newInfo.textureCopy;
+                    }
+
+                    // we need to get the shared resource HANDLE for the ID3D11Texture2D, but in order to
+                    // get that, we need to get the IDXGIResource* first.
+                    ID3D11Texture2D *texture2D = nullptr;
+                    IDXGIResource* dxgiResource = nullptr;
+                    hr = textureToMap->QueryInterface(__uuidof(IDXGIResource), (LPVOID*)&dxgiResource);
+                    if (FAILED(hr)) {
+                      m_log->error()
+                        << "RenderManagerD3D11ATW::"
+                        << "RegisterRenderBuffersInternal: Can't get the IDXGIResource for the texture resource.";
+                      setDoingOkay(false);
+                      return false;
+                    }
+
+                    // now get the shared HANDLE
+                    hr = dxgiResource->GetSharedHandle(&newInfo.sharedResourceHandle);
+                    if (FAILED(hr)) {
+                      m_log->error()
+                        << "RenderManagerD3D11ATW::"
+                        << "RegisterRenderBuffersInternal: Can't get the shared handle from the dxgiResource.";
+                      setDoingOkay(false);
+                      return false;
+                    }
+                    dxgiResource->Release(); // we don't need this anymore
+
+                                             // Get a pointer on our device to the texture we're getting
+                                             // from the caller's device.
+                    hr = atwDevice->OpenSharedResource(newInfo.sharedResourceHandle, __uuidof(ID3D11Texture2D),
+                      (LPVOID*)&texture2D);
+                    if (FAILED(hr)) {
+                      m_log->error() << "RenderManagerD3D11ATW::"
+                        << "RegisterRenderBuffersInternal: - failed to open shared resource "
+                        << "for buffer " << i;
+                      setDoingOkay(false);
+                      return false;
+                    }
```

Comment:
The original code had and if...else where both halves of the statement did a large amount of common things and the else did some additional work.  The code was changed to pull the common work out of the if...else statement and to only do the extra work for the else part.  Then the statement was inverted so that it picked out the else part of the original if...else statement.